pkg/executor/join/hash_join_base.go

```diff
@@ -218,15 +234,44 @@ type buildWorkerBase struct {
 	BuildKeyColIdx []int
 }
 
-// fetchBuildSideRows fetches all rows from build side executor, and append them
-// to e.buildSideResult.
+func syncerAdd(syncer *sync.WaitGroup) {
+	if syncer != nil {
+		syncer.Add(1)
+	}
+}
+
+func syncerDone(syncer *sync.WaitGroup) {
+	if syncer != nil {
+		syncer.Done()
+	}
+}
+
 func (w *buildWorkerBase) fetchBuildSideRows(ctx context.Context, hashJoinCtx *hashJoinCtxBase, chkCh chan<- *chunk.Chunk, errCh chan<- error, doneCh <-chan struct{}) {
-	defer close(chkCh)
-	var err error
+	w.fetchBuildSideRowsImpl(ctx, hashJoinCtx, nil, chkCh, errCh, doneCh)
+}
+
+// fetchBuildSideRowsImpl fetches all rows from build side executor, and append them
+// to e.buildSideResult.
+func (w *buildWorkerBase) fetchBuildSideRowsImpl(ctx context.Context, hashJoinCtx *hashJoinCtxBase, fetcherAndWorkerSyncer *sync.WaitGroup, chkCh chan<- *chunk.Chunk, errCh chan<- error, doneCh <-chan struct{}) {
+	var globalErr error
+	defer func() {
+		defer close(chkCh)
```

Comment:
You can define mulitple defers in a function, no need to use nested defer. For this specific case, you can move `defer close(chkCh)` to the first line of this function.

Reply:
> You can define mulitple defers in a function, no need to use nested defer. For this specific case, you can move `defer close(chkCh)` to the first line of this function.

I will put this defer outside side of the defer function